### DISCUSSION TITLE
Is getInitialProps deprecated or just not recommended?

### DISCUSSION THREAD

**Discussion by chrisdhanaraj** [vercel/next.js#17270] [Help] [answered]
Hey there!

We're rolling out some Next.js usage at my workplace, and I wanted to poke at the removal of `getInitialProps` removal from some of the main documentation paths. Our use case really falls specifically into the `getInitialProps` scenario - that is, we want to fetch once from the server on load, and then run it via the client on future routing. This is the case for things like leveraging our client cache, snappier page transitions, feature flags (we don't really need the server to get them the second time, could pass them via the route), etc. 

Anyways, trying to get an answer if the `getInitialProps` is slated for removal, or is just not recommended as much anymore for general use

**Comment by lfades** [comment, marked as answer]
`getInitialProps` **is not deprecated**. But **we do not recommend its usage** over `getStaticProps` or `getServerSideProps`. That's doesn't mean you can not use it, it only means that from our opinion you should use the new data fetching methods.

**Reply by dannycochran** [reply]
@lfades I'm curious why you'd make that recommendation though. It seems like the three methods all have valid use cases, and `getInitialProps` remains an important use case ([see this thread for others who have this use case](https://github.com/vercel/next.js/issues/13910#issuecomment-643348262)).

For `getInitialProps`, let's say I only want a bundle of translation strings and feature flags. This data is not updated very frequently (but frequently enough where I don't want to use `getStaticProps`) so it's sufficient for me to just load it on initial load of the first page a user hits. Subsequent routes would all be client side and I'd keep track of the data globally. Per [this comment](https://github.com/vercel/next.js/issues/13910#issuecomment-643348262):

```javascript
Index.getInitialProps = async ({ req }) => {
  let posts = [];
  if (req) {
    const res = await fetch("https://getpost.com/");
    posts = await res.json();
  }
  return { posts };
};
```

I think it might be worthwhile to make a clearer explanation in the documentation. Right now that message reads like deprecation.

**Reply by stegano** [reply]
> @lfades I'm curious why you'd make that recommendation though. It seems like the three methods all have valid use cases, and `getInitialProps` remains an important use case ([see this thread for others who have this use case](https://github.com/vercel/next.js/issues/13910#issuecomment-643348262)).
> 
> For `getInitialProps`, let's say I only want a bundle of translation strings and feature flags. This data is not updated very frequently (but frequently enough where I don't want to use `getStaticProps`) so it's sufficient for me to just load it on initial load of the first page a user hits. Subsequent routes would all be client side and I'd keep track of the data globally. Per [this comment](https://github.com/vercel/next.js/issues/13910#issuecomment-643348262):
> 
> ```js
> Index.getInitialProps = async ({ req }) => {
>   let posts = [];
>   if (req) {
>     const res = await fetch("https://getpost.com/");
>     posts = await res.json();
>   }
>   return { posts };
> };
> ```
> 
> I think it might be worthwhile to make a clearer explanation in the documentation. Right now that message reads like deprecation.


+1 😭

**Reply by tunesmith** [reply]
Please note that the above approach of `if (req) {` is not entirely reliable, at least not if you have multiple entry pages. If you want to load it on the first page a user hits, you may be tempted to put `getInitialProps` in a custom `_app`. However, if the user then client-side navigates to a separate page that in turn implements `getServerSideProps`, the `getInitialProps` will run on the **server side**.

In other words, the question of "Is this a client-side transition?" is different than "Is this code running on the client side?"

Our team is experimenting with looking at using `if (req.url.startsWith('/_next/data'))` for a more reliable test on whether a client-side transition is happening.

**Reply by weyert** [reply]
Can you  please give an update at later time whether `/_next/data`-check worked @tunesmith?

**Reply by tunesmith** [reply]
Checking `/_next/data` has passed our initial QA but our final product isn't live yet. I created a related feature request: https://github.com/vercel/next.js/discussions/19459

**Reply by aleksakojadinovic** [reply]
Hi there, is this discussion dead? Has anyone figured out the correct approach here? I assume the most popular answer now is 'switch to app router', however some of us are still not able to, so a definitive answer here would be very helpful.

**Comment by chrisdhanaraj** [comment]
Gotcha, thank you!
